examples/unixfs-file-cid/main.go

```diff
@@ -0,0 +1,231 @@
+	b := strings.Builder{}
+	for i := 0; i <= 100000; i++ {
+		s := strconv.Itoa(i)
+		_, err := b.WriteString(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return []byte(b.String()), nil
+}
+
+func startDataServer(ctx context.Context, h host.Host) (cid.Cid, *bsserver.Server, error) {
+	fileBytes, err := createFile0to100k()
+	if err != nil {
+		return cid.Undef, nil, err
+	}
+	fileReader := bytes.NewReader(fileBytes)
+
+	ds := dsync.MutexWrap(datastore.NewMapDatastore())
+	bs := blockstore.NewBlockstore(ds)
+	bs = blockstore.NewIdStore(bs) // handle identity multihashes, these don't require doing any actual lookups
+
+	bsrv := blockservice.New(bs, offline.Exchange(bs))
+	dsrv := merkledag.NewDAGService(bsrv)
+
+	// Create a UnixFS graph from our file, parameters described here but can be visualized at https://dag.ipfs.tech/
+	ufsImportParams := uih.DagBuilderParams{
+		Maxlinks:  uih.DefaultLinksPerBlock, // Default max of 174 links per block
+		RawLeaves: true,                     // Leave the actual file bytes untouched instead of wrapping them in a dag-pb protobuf wrapper
+		CidBuilder: cid.V1Builder{ // Use CIDv1 for all links
+			Codec:    uint64(multicodec.DagPb),
+			MhType:   uint64(multicodec.Sha2_256), // Use SHA2-256 as the hash function
+			MhLength: -1,                          // Use the default hash length for the given hash function (in this case 256 bits)
+		},
+		Dagserv: dsrv,
+		NoCopy:  false,
+	}
+	ufsBuilder, err := ufsImportParams.New(chunker.NewSizeSplitter(fileReader, chunker.DefaultBlockSize)) // Split the file up into fixed sized 256KiB chunks
+	if err != nil {
+		return cid.Undef, nil, err
+	}
+	nd, err := balanced.Layout(ufsBuilder) // Arrange the graph with a balanced layout
+	if err != nil {
+		return cid.Undef, nil, err
+	}
+
+	// Start listening on the Bitswap protocol
+	// For this example we're not leveraging any content routing (DHT, IPNI, delegated routing requests, etc.) as we know the peer we are fetching from
+	n := bsnet.NewFromIpfsHost(h, routinghelpers.Null{})
+	bswap := bsserver.New(ctx, n, bs)
+	n.Start(bswap)
+	return nd.Cid(), bswap, nil
+}
+
+func runClient(ctx context.Context, h host.Host, c cid.Cid, targetPeer string) ([]byte, error) {
+	n := bsnet.NewFromIpfsHost(h, routinghelpers.Null{})
+	bswap := bsclient.New(ctx, n, blockstore.NewBlockstore(datastore.NewNullDatastore()))
+	n.Start(bswap)
+	defer bswap.Close()
+
+	// Turn the targetPeer into a multiaddr.
+	maddr, err := multiaddr.NewMultiaddr(targetPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract the peer ID from the multiaddr.
+	info, err := peer.AddrInfoFromP2pAddr(maddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Directly connect to the peer that we know has the content
+	// Generally this peer will come from whatever content routing system is provided, however go-bitswap will also
+	// ask peers it is connected to for content so this will work
+	if err := h.Connect(ctx, *info); err != nil {
+		return nil, err
+	}
```

Comment:
Probably the "safer" way to do this if you were running this for real would be to create a content router that always returned the the target peer had the data. This is because if the remote peer disconnects from you (or you from them) you have no way to re-establish the connection. Not sure if it's worth adding more complexity to the code here though.

Ideally any API changes around sessions should make doing this less painful, although if we needed a special content router that wouldn't be the end of the world (some food for thought @Jorropo).

Reply:
you could mock a content routing that always return you `info` but that would be very confusing to new comers, not the goal of an example, that looks fine to me. I guess the only better way would be to actually use the dht.